[Foundation] Handle null for Optional<T> in JSONDecoder and PropertyListDecoder.

dingobye opened the pull request:
This patch makes JSONDecoder and PropertyListDecoder accept null for Optional<T> values when using SingleValueDecodingContainer.

Resolves [SR-7404](https://bugs.swift.org/browse/SR-7404).

dingobye committed:
[Foundation] Decode null for Optional<T> in JSONDecoder.

Make JSONDecoder accept null for Optional<T> values
via SingleValueDecodingContainer.

Resolves: SR-7404.

dingobye commented:
@itaiferber Would you help review, please? Thank you.

itaiferber commented:
Thanks for putting this together! The approach I had in mind for resolving this was a little bit different (I was thinking of actually excising all of the optional returns from the `unbox` methods and handling optionality differently), but this is a concise method too (which makes use of the fact that `unbox` returns the type wrapped in another layer of optionality).

Let's test, but I'm not sure whether we should be using `_isOptional` to achieve this. @airspeedswift Thoughts on this usage?

itaiferber commented:
@swift-ci Please test macOS

dingobye commented:
Hi Itai, thank you for the feedback!

Agree that for efficiency considerations, if `_isOptional` is expensive then we should avoid such usage. I am not sure how expensive it is, though. Looks like it eventually calls [this C++ function](https://github.com/apple/swift/blob/master/stdlib/public/runtime/Casting.cpp#L3086), which appear to be performant...

If we don't want to have `_isOptional` here, I guess we may use `T.self as? ExpressibleByNilLiteral.Type` as an alternative to check for optionality. What do you think?

airspeedswift commented:
Foundation should not be using underscored methods that are defined inside the standard library (except in a handful of cases like bridging where they are there explicitly for Foundation).

Maybe try overloading, which should work unless you're in a fully generic context:

```swift
public func decode<T : Decodable>(_ type: T?.Type) {
  print("it was optional")
}

public func decode<T : Decodable>(_ type: T.Type) {
  print("it wasn't optional")
}

decode((Int?).self)
decode((Int).self)
```

airspeedswift commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -2005,7 +2005,9 @@ extension _JSONDecoder : SingleValueDecodingContainer {
 
     public func decode<T : Decodable>(_ type: T.Type) throws -> T {
-        try expectNonNull(type)
+        if !_isOptional(type) {
```
This is for std lib internal use only. It will probably be made private in Swift 5 (now that we have `@usableFromInline`)

itaiferber replied:
@millenomi Suggested taking a look at `_Unwrappable`, which does something similar and is used in bridging. Alternatively, we could introduce the same type of protocol locally here and conform similarly, but we've run into issues with doing something like this in #16238; thoughts?

We can't take the overloading approach because that candidate wouldn't get called — the interface here is fully generic (the exposed interface to calling these methods is the existential `SingleValueDecodingContainer`, not the concrete type).

itaiferber replied:
Looks like `_Unwrappable` isn't public, though, so I guess we don't have much of a choice here.

itaiferber commented:
@dingobye That has the potential for catching other types which are `ExpressibleByNilLiteral`, which we'd like to avoid. An alternative approach here is defining a `fileprivate` protocol locally which only `Optional` conforms to, then casting to that metatype.

As mentioned in my comment to @airspeedswift, we had trouble with this approach in #16238, but if we can figure out the underlying issue there, we could make it work.

dingobye committed:
[Foundation] Decode null for Optional<T> in JSONDecoder.

Avoid using _isOptional to check for optionality.
Instead, make a fileprivate empty protocol, which
is exclusively conformed to by Optional.

dingobye commented:
Thank you for the informative feedback! 

I rebased and updated the code to avoid using underscore functions designed for stdlib internal use. As @itaiferber  suggested, a fileprivate empty protocol, which is conformed to exclusively by Optional, is introduced to achieve this.

The issue in #16238  looks so complicated that I don't have any clue where to focus on. Fortunately, the case in this patch is simpler.

Would you help fire a test again and see if it succeeds on CI server, please? Many thanks!

jrose-apple commented:
@swift-ci Please test macOS

dingobye commented:
Thank you, Jordan!

Swift Test OS X Platform [completed successfully](https://ci.swift.org/job/swift-PR-osx/6101/); however @swift-ci said "Build failed", and the failure seems to be with an invalid sha. Feel confused...

jrose-apple commented:
That's an artifact of the GitHub/Jenkins plugin; there's a bug where it tries to build the wrong commit first if you do a force push. We've got it set up so that the build will automatically restart in that case, which is what happened here.

itaiferber commented on stdlib/public/SDK/Foundation/JSONEncoder.swift:
```diff
@@ -2453,3 +2456,18 @@ fileprivate extension EncodingError {
+
+@inline(__always)
+fileprivate func _isOptionalType<T>(_ type: T.Type) -> Bool {
+    return type as? _Unwrappable.Type != nil
```
This can be more easily written as `type is _Unwrappable.Type`, but looks reasonable to me otherwise overall.

dingobye replied:
Fixed.

itaiferber reviewed (commented):
Looks reasonable to me — I'll try to make another pass on this to see if we can improve on this approach in any way.

While we're at it, can we reasonable factor this out (into maybe an `internal` protocol + function) so we can apply the same to `PropertyListDecoder` as well?

dingobye committed:
[Foundation] Decode null for Optional<T> in JSONDecoder and PropertyListDecoder.

Make optionality checker function internal; and apply
the same scheme to PropertyListDecoder to accept null
when decoding Optional values with SingleValueDecodingContainer.

itaiferber commented:
@dingobye So sorry that this slipped through my fingers. The latest changes look reasonable to me. An alternative location for the optionality stuff could also be Foundation's own `Codable.swift` file, but this should be fine.

Let's test this.

itaiferber commented:
@swift-ci Please test macOS

itaiferber commented:
@swift-ci Please smoke test Linux

itaiferber commented:
@airspeedswift Do you have any outstanding remarks on this? Is the latest approach reasonable?

CodaFi commented:
@itaiferber Do we still want to take this change - it seems like the thing that covers the SR is `decodeIfPresent`.

itaiferber commented:
@CodaFi Thanks for resurfacing this. I'm no longer at Apple so I think @parkera should weigh in on a decision.

CodaFi commented:
Good idea.  @millenomi may also have a good opinion here.

futurejones commented:
Can one of the admins verify this patch?

shahmishal commented:
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))
